src/goto-instrument/nondet_static.cpp

```diff
@@ -136,3 +140,68 @@ void nondet_static(goto_modelt &goto_model)
   const namespacet ns(goto_model.symbol_table);
   nondet_static(ns, goto_model.goto_functions);
 }
+
+/// Second main entry point of the module. Nondeterministically initializes
+/// global scope variables, except for constants (such as string literals, final
+/// fields), internal variables (such as CPROVER and symex variables,
+/// language specific internal variables) and variables passed to except_value.
+/// \param [out] goto_model: Existing goto-model to be updated.
+/// \param except_values: list of symbol names that should not be updated.
+void nondet_static(
+  goto_modelt &goto_model,
+  const optionst::value_listt &except_values)
+{
+  const namespacet ns(goto_model.symbol_table);
+  std::set<std::string> to_exclude;
+
+  for(optionst::value_listt::const_iterator i = except_values.begin();
+      i != except_values.end();
+      ++i)
+  {
+    bool prefix_found = false;
+    for(size_t j = 0; j < i->length(); ++j)
+    {
+      if((*i)[j] == ':')
+      {
+        prefix_found = true;
+        to_exclude.insert(*i);
+        if(i->length() > 2 && (*i)[0] == '.' && (*i)[1] == '/')
+        {
+          to_exclude.insert(i->substr(2, i->length() - 2));
+        }
+        else
+        {
+          to_exclude.insert("./" + *i);
+        }
+        break;
+      }
+    }
+    if(!prefix_found)
+    {
+      irep_idt symbol_name(*i);
+      symbolt lookup_results = ns.lookup(symbol_name);
+      to_exclude.insert(
+        id2string(lookup_results.location.get_file()) + ":" + *i);
+    }
+  }
+
+  symbol_tablet symbol_table = ns.get_symbol_table();
```

Comment:
Why is a copy taken?

Reply:
Because namespacet has a const symbol_table.
If I take a const reference the iterators are also const and I cannot call get_writable_symbol on the iterator. Which makes sense from my point of view that a const symbol_table must not be modified. Anyhow, I want to modify it. Therefore, I copy it and then modify the copy.

Eventually, I have overseen a possibility to this without the copy. But I don’t know how to get a symbol_tablet::iteratort instead of the symbolst::const_iterator if I have a const reference to a symbol table.

Any advice?

Reply:
You don't really need to go through `ns`, because the namespace itself you only constructed earlier from `goto_model.symbol_table` - I think you could just use that symbol table directly?